Improved partial hydration code

gregod opened the pull request:
This PR improves the partial hydration code sent to the client. They mostly aim at reducing code size, but 272b131 should improve concurrency in loading props. 

There are 4 commits in this pull requests with increasing "aggressiveness" in changing the current implementation.  Feel free to only pick a subset if you disagree with my suggestions.

I haven't tested against real compressed props, as i currently have no data that seems to trigger it. But everything is passed transparently through the $ejs function.  Could you confirm this works ok and then update the snapshot tests? (Or let me know and i will revise this PR)

gregod commented on src/partialHydration/hydrateComponents.ts:
```diff
@@ -159,17 +149,17 @@ export default (page: Page) => {
 
     if (component.hydrateOptions.loading === 'eager') {
-      eagerString += `['${component.name}','${component.client.replace(`${relPrefix}/svelte/components/`, '')}', ${
+      eagerString += `'${component.name}' : { 'component' : '${component.client.replace(`${relPrefix}/svelte/components/`, '')}', 'props' : ${
```
We could shave off a few more bytes by turning the named attributes into an array again, but i don't thing the clarity loss is worth it.

nickreese replied:
I agree

nickreese commented:
Great work. I’m on mobile, on vacation, with a bad case of Jetlag but things look good. 

Only question is if we should be reusing the IO as eager loaded components get called separately via the same function.  

The messy code you refactored was partly so complex because I was telling myself a story that it would be harder for people to scrape if they couldn’t grok it. 

Your proposal is much better. Good catch on the await.

gregod commented:
I think the proper differentiation between eager and lazy got lost somewhere in the v1.5 rewrite. Even before this PR the only difference was that lazy got the wrap into requestIdleCallback. The Interaction Observer was always applied.
I will re-implement this functionality sometime during the weekend and push an update in this PR.

The HTTP requests would be trivial to find anyway, so might as well be open there. But now I understand the preference for prop compression a bit better ;) 

With all the templating of js in js and the added functionality the code generation code has turned into quite the mess. Depending of how much time i have i will separately submit a PR that refactors the whole section after these changes here have merged.

gregod commented:
I have restored the eager loading in 960eb21 and added an additional optimization to elide the lazy loading code if all components on the page are eager loading. 

On my test page with two components this leads to a 8% decrease in hydration code after minification ( -35% if only eager components are present in the page) compared to the current master, while being more correct in terms of eager loading and concurrency

nickreese commented:
@gregod Great work. Will review in detail tomorrow when at the laptop. One thing I noticed and wanted to discuss was the thought of lazy loading the props vs not.

Currently I'm in Spain and only have a very unstable 4g connection tethered to a prepaid mobile phone. If the props were lazy loaded with the promise all instead of loaded when the `requestIdleCallback` fired, it could take 4-5s for the component to mount. Any ideas on how we could address this situation?

It seems firing the network requests off and getting the props before they are needed is a smart move in most situations?

gregod commented:
Good point about firing the requests off early. I was thinking more in the direction of reducing the data transmitted, especially on slow/metered connections. But yeah, if we rely on the intersection observer the request might not make it in time.

So maybe return to requesting props and component eagerly and only defer the more computation heavy decompression and hydration itself using the IO?

I see one alternate implementation option:
We set the component preload attribute to true by default (thus injecting the preload html tag) and let the browser handle it.  This has the advantage that the request will be sent with a low priority (wont delay any other asset requests) and behaviour can be modified by the browser / user settings. For truly lazy components one could just explicitly set the attribute to false. The disadvantage is that there are differences in preloading for modules in chromium browsers (~50% of US users https://caniuse.com/link-rel-modulepreload) where the current rel=preload tag leads to duplicate requests.
There is already a similar comment in https://github.com/Elderjs/elderjs/blob/2f3f8c69d421ab490ae776eee157bec4d0050bd4/src/partialHydration/hydrateComponents.ts#L185-L186
about these preload methods.

nickreese commented:
@gregod It seems based on my reading that `preload`/`modulepreload` isn't 100% sorted. 

Some sources are saying that `preload` is high priority. I seems the `import` within the `requestIdleCallback` may be the most efficient... though I'd like to let the browser do the handling of the fetching if we could. @lukeed thoughts here?

gregod commented:
I must have mixed up prefetch and preload. The priority is indeed high for preload.
Unfortunately there is no moduleprefetch. Together with the inconsistent browser behavior this looses against the JS solution. Maybe this could be revisited once the dust settles.

I have pushed an update that does import both components and props right away and only delays the decompression and hydration via the IO.

gregod commented:
The web vitals look as expected. FCP is right after all "eager" components and props have been fetched, but before the lazy components. If i remember correctly this is the same behavior as shown with the preload tags. Im not familiar with the impact of web vitals on SEO.

I noticed that all components additionally depend on the index_[hash].js from svelte. If we revisit the preloading this should be preloaded as well.

lukeed commented:
I think a balance of everything listed here would be to restart from the current Elder implementation, using IO to add a `link[rel=prefetch]` for any visible component, and then keeping the `import()` call inside the rIC whenever a component needs to be initialized.

This way the network activity is still tied to visibility but is already done once it's time for work to happen. The two-phase approach could make room for a "when to hydrate" modifier – something like `<div hydrate:visible={ props } />` vs `hydrate:idle` vs `hydrate:hover` vs ... .

nickreese commented:
@gregod Just checking in here. It seems we've basically got what @lukeed is suggesting he is just proposing a minor improvement of adding IO calls to add `prefetch` to the components that are visible. 

This basically streamlines the process and gets us what appears to be the best of both worlds.

Thoughts here?

gregod commented:
I'm not 100% understanding their proposal. The need to initialize a component is directly in the IO.

So this is in the current master
a) `Page Load -> Idle Callback -> Promise Fetch & Await Props ->    Wait for IO Visibility -> Await Component -> Hydrate`
The await is blocking too early, so the more correct version would be:
b) `Page Load -> Idle Callback -> Promise Fetch Props ->    Wait for IO Visibility -> Fetch Component -> Await Component & Props -> Hydrate`

Why would we preload the props and not the component? This depends of course on the ratio of component vs props size, but unless  the component is huge i don't see it. Further the component is most likely to be in the cache on a visit to a second page.

The current version of this PR does the following:
c) `Page Load -> Idle Callback -> Promise Fetch Props & Component ->   Wait for IO Visibility -> Await Props & Component -> Hydrate`
The initial proposal in the PR did this:
d) `Page Load -> Idle Callback -> Wait for IO Visibility -> Promise Fetch Props & Component  -> Await Props & Component -> Hydrate`
If the preload prop is activated the following would happen with d).
e) `Page Load -> Prefetch/Preload through static meta tag -> Idle Callback -> Wait for IO Visibility -> Promise Fetch Props & Component  -> Await Props & Component -> Hydrate`
Still has the issue of not working in Chrome/Edge (with prefetch) and causing duplicate requests.

Is  this what @lukeed are suggesting?
f) `Page Load -> Idle Callback ->  Wait for IO Visibility -> Add Prefetch Tag -> Promise Fetch Props & Component (likely still in flight) -> Await Props & Component -> Hydrate`

I do not see the difference unless the hydration is further delayed as suggested.
g) `Page Load -> Idle Callback ->  Wait for IO Visibility -> Add Prefetch Tag -> Wait for Hover/Interaction -> Promise Fetch Props & Component (likely completed) -> Await Props & Component -> Hydrate`

nickreese commented:
@gregod Sorry for the delay. In the middle moving from the US to Europe and haven't had time at the laptop.

I think @lukeed's suggestion essentially amounts to the current PR, but adding a prefetch for components that are in the viewbox on page load. 

I really like the current status of the PR and am good with merging it. 

I do like the idea of some sort of visualization when on a slow connection as I'm experiencing 2-3g a lot in my travels but I think it may be best to revisit it in another PR or once we've hashed out the ideal feature set.

lukeed commented:
My suggestion was just to add `link[prefetch]`s immediately and keep deferring all `import()`s until after the rIC is ready. Not a massive difference, but moves/keeps the network request initialization separate from the `import()`, which is what kicks off the CPU workload

gregod commented:
@lukeed Ah! Yes this makes sense. Maybe the easiest way would be to add this as option on the component, parallel to the "preload" option which already exists? This would allow for easy customization of which components are critical and which are not.  (The issue that chromium based browsers have taken a different path with modulepreload unfortunately remains). Should we default to true or false?

I would be open to adding this prefetch setting and then this PR should be ready. We could continue the other ideas in a separate discussion.

gregod commented:
I finally found some time to revisit this again. 

I have added the prefetch option as default (using a `noPrefetch` override, so its set by default). As the modulepreload / preload question seems to be still open and JSON should be lighter on the parser i have moved the props into a .json file instead of an module. This way we have proper prefetching for the props file no matter how this discussion will settle.

While testing this i noticed that my current chromium does correctly reuse the prefetched js modules! (Not sure if this was always the case. When testing "preload" before there were duplicate requests)  However having them as raw json might still have some performance benefits so i'm still suggesting to use the json for props.

The prefetching works so well that I'm also suggesting a opinionated change in 9faff64 to only do the real fetch within the intersection observer for "lazy" components. To give the browser a little head start for the request and hydration on slow connections i added an offset of 200px to the intersection observer.

This should give us the best of both worlds: An early request for all resources in the background and parsing and hydration only once the viewport is close to the component for "lazy" components. The "eager" components get the benefit of prefetching, and we skip the intersection observer to initialize them right away.

Let me know what you think. 
I would be also fine with applying only a subset of the commits if you have a different opinion.

nickreese commented:
Looks great on mobile, will fully review when in the office tomorrow.

nickreese commented:
OK, got a good chance to review the code. I was having an obscure bug (user error) that blocked me when I tried last week.

This looks great and I agree with the opinionated change as it seems like a good balance of all of the concerns.

In my testing, somewhere before this patch, I realized we've dropped support for customizing the `rootMargin` and `threshold` when we made the move to a single `IO`:

`hydrate-options={{ rootMargin: '500px', threshold: 0 }}`

I think the condensed code is worth the loss of configuration as the tradeoff of complexity just isn't worth it as the user may as well `eager` load given a large enough `rootMargin`.

Great work, I'll get the docs updated with `noPrefetch` support and ship this.
